### PR TITLE
MULE-12716: Remove registries from API

### PR DESCRIPTION
**Pull request by elrodro83**
* Final round

Created issues to cleanup some dirty code introduced with this change:
* https://www.mulesoft.org/jira/browse/MULE-13626
* https://www.mulesoft.org/jira/browse/MULE-13627

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/internal/context/MuleContextWithRegistries.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * @since 4.0
+ */
+// TODO MULE-13627 Remove downcasts to this interface
+public interface MuleContextWithRegistries extends MuleContext {
+
+  MuleRegistry getRegistry();
+
+  /**
+   * @deprecated as of 3.7.0. This will be removed in Mule 4.0
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/internal/context/MuleContextWithRegistries.java]: Update javadocs, it makes no sense to reference 3.7.0 and this probably wont be  removed on 4.0 :)

File: core/src/main/java/org/mule/runtime/core/internal/execution/FlowProcessingPhase.java [hunk 2]
```diff
@@ -38,6 +40,12 @@
 
   protected transient Logger logger = LoggerFactory.getLogger(getClass());
 
+  private Registry registry;
+
+  public FlowProcessingPhase(Registry registry) {
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/internal/execution/FlowProcessingPhase.java]: Add javadoc

File: core/src/main/java/org/mule/runtime/core/internal/execution/MuleMessageProcessingManager.java [hunk 3]
```diff
@@ -34,26 +38,34 @@
...
 
-  @Override
+  @Inject
   public void setMuleContext(MuleContext context) {
     this.muleContext = context;
   }
 
+  @Inject
+  public void setRegistryMessageProcessPhases(Optional<Collection<MessageProcessPhase>> registryMessageProcessPhases) {
+    // This needs to be an Optional due to https://jira.spring.io/browse/SPR-15338
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/internal/execution/MuleMessageProcessingManager.java]: Does this need a TODO to change it once the referenced issue is fixed?

File: core/src/main/java/org/mule/runtime/core/internal/processor/InvokerMessageProcessor.java [hunk 4]
```diff
@@ -106,26 +106,27 @@ protected void resolveMethodToInvoke() throws InitialisationException {
...
-      LOGGER.debug(String.format("Initialised %s to use method: '%s'", this, method));
+      LOGGER.debug(format("Initialised %s to use method: '%s'", this, method));
     }
   }
 
   protected void lookupObjectInstance() throws InitialisationException {
     if (LOGGER.isDebugEnabled()) {
-      LOGGER.debug(String.format("No object instance speciedied.  Looking up single instance of type %s in mule registry",
-                                 objectType));
+      LOGGER.debug(format("No object instance speciedied.  Looking up single instance of type %s in mule registry",
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/internal/processor/InvokerMessageProcessor.java]: speciedied > specified?

File: tests/unit/src/main/java/org/mule/tck/MuleTestUtils.java [hunk 5]
```diff
@@ -55,19 +63,36 @@ public static Injector spyInjector(MuleContext muleContext) {
...
     // Use direct processing strategy given flow used in test event is not used for processing.
-    final Flow flow = builder(APPLE_FLOW, context).withDirectProcessingStrategyFactory().build();
-    flow.setAnnotations(singletonMap(LOCATION_KEY, fromSingleComponent(APPLE_FLOW)));
-    if (context.getRegistry() != null) {
-      context.getRegistry().registerFlowConstruct(flow);
-    }
+    return createFlow(context, APPLE_FLOW);
+  }
 
+  public static Flow createFlow(MuleContext context, String flowName) throws MuleException {
```
Comment (pablokraan) [inline tests/unit/src/main/java/org/mule/tck/MuleTestUtils.java]: Add javadoc

File: tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java [hunk 6]
```diff
@@ -75,13 +90,76 @@ public static MuleContext mockMuleContext() {
...
+                dependencyType = (Class<?>) type;
+              }
+              nullToOptional = true;
+            }
+
+            if (MuleContext.class.isAssignableFrom(dependencyType)) {
+              try {
+                method.invoke(object, nullToOptional ? of(muleContext) : muleContext);
+              } catch (Exception e) {
+                throw new RuntimeException(format("Could not inject dependency on method %s of type %s", method.getName(),
```
Comment (pablokraan) [inline tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java]: Extract exception creation into an utility method

File: tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java [hunk 7]
```diff
@@ -134,8 +216,12 @@ public static void registerIntoMockContext(MuleContext context, String key, Obje
    * @return a basic event builder with its context already set.
    */
   public static <B extends CoreEvent.Builder> B eventBuilder() throws MuleException {
-    FlowConstruct flowConstruct = getTestFlow(mockContextWithServices());
-    return (B) InternalEvent.builder(create(flowConstruct, TEST_CONNECTOR_LOCATION));
+    return eventBuilder(mockContextWithServices());
+  }
+
+  public static <B extends CoreEvent.Builder> B eventBuilder(MuleContext muleContext) throws MuleException {
```
Comment (pablokraan) [inline tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java]: Add javadoc

File: core/src/main/java/org/mule/runtime/core/internal/execution/MuleMessageProcessingManager.java (hunk 3 above)
Comment (elrodro83) [reply core/src/main/java/org/mule/runtime/core/internal/execution/MuleMessageProcessingManager.java]: No, it won;t be fixed till Spring 5 at least. It's just that at first i tried without the `Optional` and it failed and i investigated, Just added the comment to discourage anyone to try to simplify this.

File: tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java (hunk 6 above)
Comment (elrodro83) [reply tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java]: the messages are differnet

**Pull request merged**